apollo-router/src/axum_factory/compression/mod.rs

```diff
@@ -0,0 +1,186 @@
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+use tower::BoxError;
+
+use self::codec::BrotliEncoder;
+use self::codec::DeflateEncoder;
+use self::codec::Encode;
+use self::codec::GzipEncoder;
+use self::codec::ZstdEncoder;
+use self::util::PartialBuffer;
+
+pub(crate) mod codec;
+pub(crate) mod unshared;
+pub(crate) mod util;
+
+pub(crate) enum Compressor {
+    Deflate(DeflateEncoder),
+    Gzip(GzipEncoder),
+    Brotli(BrotliEncoder),
+    Zstd(ZstdEncoder),
+}
+
+//FIXME: we should call finish at the end
+impl Compressor {
+    pub(crate) fn new<'a, It: 'a>(it: It) -> Option<Self>
+    where
+        It: Iterator<Item = &'a str>,
+    {
+        for s in it {
+            match s {
+                "gzip" => return Some(Compressor::Gzip(GzipEncoder::new(Compression::fast()))),
+                "deflate" => {
+                    return Some(Compressor::Deflate(
+                        DeflateEncoder::new(Compression::fast()),
+                    ))
+                }
+                // FIXME: find the "fast" brotli encoder params
+                "br" => {
+                    return Some(Compressor::Brotli(BrotliEncoder::new(
+                        BrotliEncoderParams::default(),
+                    )))
+                }
+                "zstd" => {
+                    return Some(Compressor::Zstd(ZstdEncoder::new(zstd_safe::min_c_level())))
+                }
+                _ => {}
+            }
+        }
+        None
+    }
+
+    pub(crate) fn content_encoding(&self) -> &'static str {
+        match self {
+            Compressor::Deflate(_) => "deflate",
+            Compressor::Gzip(_) => "gzip",
+            Compressor::Brotli(_) => "br",
+            Compressor::Zstd(_) => "zstd",
+        }
+    }
+
+    pub(crate) fn process(
+        mut self,
+        mut stream: hyper::Body,
+    ) -> impl Stream<Item = Result<Bytes, BoxError>>
+where {
+        let (tx, rx) = mpsc::channel(10);
+
+        tokio::task::spawn(async move {
+            while let Some(data) = stream.next().await {
+                match data {
+                    Err(e) => {
+                        if let Err(_) = tx.send(Err(e.into())).await {
+                            return;
+                        }
+                    }
+                    Ok(data) => {
+                        let mut buf = BytesMut::zeroed(1024);
+                        let mut written = 0usize;
+
+                        let mut partial_input = PartialBuffer::new(&*data);
+                        loop {
+                            let mut partial_output = PartialBuffer::new(&mut buf);
+                            partial_output.advance(written);
+
+                            match self.encode(&mut partial_input, &mut partial_output) {
+                                Err(e) => panic!("{e:?}"),
+                                Ok(()) => {}
+                            }
+
+                            //let read = partial_input.written().len();
```

Comment:
```suggestion
```

Reply:
fixed in [0fb0ba9](https://github.com/apollographql/router/pull/2986/commits/0fb0ba985404c3652c257bed12044f955b154d53)